src/main/java/com/yahoo/bullet/pubsub/memory/MemoryPublisher.java

```diff
@@ -0,0 +1,95 @@
+/*
+ *  Copyright 2017, Yahoo Inc.
+ *  Licensed under the terms of the Apache License, Version 2.0.
+ *  See the LICENSE file associated with the project for terms.
+ */
+package com.yahoo.bullet.pubsub.memory;
+
+import com.yahoo.bullet.common.BulletConfig;
+import com.yahoo.bullet.pubsub.Metadata;
+import com.yahoo.bullet.pubsub.PubSubException;
+import com.yahoo.bullet.pubsub.PubSubMessage;
+import com.yahoo.bullet.pubsub.Publisher;
+import lombok.extern.slf4j.Slf4j;
+import org.asynchttpclient.AsyncHttpClient;
+import org.asynchttpclient.Response;
+
+import java.io.IOException;
+import java.util.function.Consumer;
+
+@Slf4j
+public abstract class MemoryPublisher implements Publisher {
+    protected MemoryPubSubConfig config;
+    protected AsyncHttpClient client;
+
+    /**
+     * Create a MemoryQueryPublisher from a {@link BulletConfig} and a {@link AsyncHttpClient}.
+     *
+     * @param config The config.
+     * @param client The client.
+     */
+    public MemoryPublisher(BulletConfig config, AsyncHttpClient client) {
+        this.config = new MemoryPubSubConfig(config);
+        this.client = client;
+    }
+
+    @Override
+    public void send(String id, String content) throws PubSubException {
+        send(new PubSubMessage(id, content, new Metadata()));
+    }
+
+    @Override
+    public void close() {
+        try {
+            client.close();
+        } catch (IOException e) {
+            log.warn("Caught exception when closing AsyncHttpClient: " + e);
+        }
+    }
+
+    /**
+     * Send the PubSubMessage to the given uri.
+     *
+     * @param uri The uri to which to send the message.
+     * @param message The message to send.
+     */
+    protected void send(String uri, PubSubMessage message) {
+        log.debug("Sending message: " + message + " to uri: " + uri);
+        client.preparePost(uri)
+              .setBody(message.asJSON())
+              .setHeader("content-type", "text/plain")
+              .execute()
+              .toCompletableFuture()
+              .exceptionally(this::handleException)
+              .thenAcceptAsync(createResponseConsumer(message.getId()));
+    }
+
+    /**
+     * Send the PubSub message. The MemoryQueryPublisher subclass will put the response uri in the MetaData of the
+     * PubSubMessage. The MemoryResultPublisher subclass will extract the uri to determine the host to which it should
+     * send the Response.
+     *
+     * @param message The message to send.
+     * @throws PubSubException
+     */
+    @Override
+    public abstract void send(PubSubMessage message) throws PubSubException;
```

Comment:
I am not sure we should re-declare this method here since it's already declared in super class.

Reply:
I like to have it here just for clarity (comments explaining what the child classes do).   I could go either way though I guess.  Let me know and we can talk about it offline.